### PR TITLE
ReadControl and WriteConrol have SharedPointers to their Containers

### PR DESCRIPTION
**Pull request by JonathanMeans**

### CODE REVIEW

File: six/modules/c++/six.sicd/tests/test_read_sicd_with_extra_des.cpp [hunk 1]
```diff
@@ -95,11 +95,12 @@ std::auto_ptr<TempFile> createNITFFromXML(const std::string& xmlPathname)
     std::vector<six::UByte> bandData(
             generateBandData(*data));
 
-    six::Container container(six::DataType::COMPLEX);
-    container.addData(dynamic_cast<six::Data*>(data.release()));
+    mem::SharedPtr<six::Container> container(new six::Container(
+            six::DataType::COMPLEX));
+    container->addData(dynamic_cast<six::Data*>(data.release()));
```
Comment (asylvest) [inline six/modules/c++/six.sicd/tests/test_read_sicd_with_extra_des.cpp]: Unrelated to these changes but the `dynamic_cast` isn't needed is it?

File: six/modules/c++/six.sidd/include/six/sidd/GeoTIFFReadControl.h [hunk 2]
```diff
@@ -44,7 +44,6 @@ class GeoTIFFReadControl : public ReadControl
     //!  Destructor
     virtual ~GeoTIFFReadControl()
     {
-        delete mContainer;
```
Comment (asylvest) [inline six/modules/c++/six.sidd/include/six/sidd/GeoTIFFReadControl.h]: Go ahead and remove the destructor altogether since it's empty now

File: six/modules/c++/six.sidd/include/six/sidd/GeoTIFFWriteControl.h [hunk 3]
```diff
@@ -59,12 +59,12 @@ class GeoTIFFWriteControl : public WriteControl
 public:
     GeoTIFFWriteControl();
     virtual ~GeoTIFFWriteControl() {}
-    
+
```
Comment (asylvest) [inline six/modules/c++/six.sidd/include/six/sidd/GeoTIFFWriteControl.h]: Unrelated but go ahead and remove this destructor too... with good RAII design, we only expect destructors when something funky is going on, so having noop ones just make us pause for a sec for no reason

File: six/modules/c++/six.sidd/source/GeoTIFFWriteControl.cpp [hunk 4]
```diff
@@ -39,7 +39,7 @@ GeoTIFFWriteControl::GeoTIFFWriteControl()
                                                     tiff::Const::Type::ASCII,
                                                     Constants::GT_XML_TAG);
 }
-void GeoTIFFWriteControl::initialize(Container* container)
+void GeoTIFFWriteControl::initialize(mem::SharedPtr<Container> container)
```
Comment (asylvest) [inline six/modules/c++/six.sidd/source/GeoTIFFWriteControl.cpp]: Do this under a separate pull request but can you add unittests showing reading and writing SIDD GeoTIFFs works if there aren't some already?

File: six/modules/c++/six/include/six/NITFWriteControl.h [hunk 5]
```diff
@@ -78,7 +78,7 @@ class NITFWriteControl : public WriteControl
     // Get the container that we were initialized from (const)
     const Container* getContainer() const
     {
-        return mContainer;
+        return mContainer.get();
```
Comment (asylvest) [inline six/modules/c++/six/include/six/NITFWriteControl.h]: Shouldn't this signature change to match the ReadControl and return a shared pointer?  Also probably move it to the base class for consistency.

File: six/modules/c++/six/include/six/ReadControl.h [hunk 6]
```diff
@@ -89,19 +89,17 @@ class ReadControl
                       const std::vector<std::string>& schemaPaths) = 0;
 
     /*!
-     *  Get a const pointer to the current container.  This
-     *  container is not allocated, nor deleted by this class
+     *  Get a const pointer to the current container.
```
Comment (asylvest) [inline six/modules/c++/six/include/six/ReadControl.h]: This and other comments should be updated to reflect a shared pointer is returned

File: six/modules/c++/six/include/six/WriteControl.h [hunk 7]
```diff
@@ -175,12 +176,11 @@ class WriteControl
...
 
     /*!
-     *  const pointer to Container.  This object is not owned by the
-     *  WriteControl
+     *  const pointer to Container.
      */
     const Container* getContainer()
     {
-        return mContainer;
+        return mContainer.get();
```
Comment (asylvest) [inline six/modules/c++/six/include/six/WriteControl.h]: Ditto on changing this to return a shared pointer

File: six/modules/c++/six/source/NITFWriteControl.cpp [hunk 8]
```diff
@@ -98,12 +98,11 @@ NITFWriteControl::NITFWriteControl()
...
 }
 
-void NITFWriteControl::initialize(Container* container)
+void NITFWriteControl::initialize(mem::SharedPtr<Container> container)
 {
     // Clean up
-    // NOTE  We do not own the container, so we don't delete 'mContainer' here
     mInfos.clear();
-
+    mContainer.reset();
```
Comment (asylvest) [inline six/modules/c++/six/source/NITFWriteControl.cpp]: This line isn't needed... the assignment on the next line will take care of everything

File: six/projects/csm/source/SICDSensorModel.cpp [hunk 9]
```diff
@@ -89,7 +89,7 @@ void SICDSensorModel::initializeFromFile(const std::string& pathname)
         reader.setXMLControlRegistry(&xmlRegistry);
         reader.load(pathname, mSchemaDirs);
 
-        six::Container* const container = reader.getContainer();
+        mem::SharedPtr<six::Container> const container = reader.getContainer();
         if (container->getDataType() != six::DataType::COMPLEX ||
```
Comment (asylvest) [inline six/projects/csm/source/SICDSensorModel.cpp]: `const` should move to the front of the line now

File: six/projects/csm/source/SIDDSensorModel.cpp [hunk 10]
```diff
@@ -116,7 +116,7 @@ void SIDDSensorModel::initializeFromFile(const std::string& pathname,
         // For multi-image SIDDs, all the SIDD DESs will appear first (in the
         // case where SICD DESs are also present), so we just have to grab out
         // the Nth Data object
-        six::Container* const container = reader.getContainer();
+        mem::SharedPtr<six::Container> const container = reader.getContainer();
```
Comment (asylvest) [inline six/projects/csm/source/SIDDSensorModel.cpp]: `const` should move to the front of the line now

File: six/modules/c++/six/include/six/ReadControl.h [hunk 11]
```diff
@@ -89,19 +89,17 @@ class ReadControl
...
      */
-    const Container* getContainer() const
+    mem::SharedPtr<const Container> getContainer() const
```
Comment (asylvest) [inline six/modules/c++/six/include/six/ReadControl.h]: Is the Solaris compiler smart enough to convert from a non-const shared pointer to a const shared pointer?  Sometimes it's dumb about stuff like that.

Comment (JonathanMeans) [inline six/modules/c++/six/include/six/ReadControl.h]: Solaris is fine with this. Everything else is fixed.

File: six/modules/c++/six/include/six/WriteControl.h [hunk 12]
```diff
@@ -175,10 +176,9 @@ class WriteControl
     }
 
     /*!
-     *  const pointer to Container.  This object is not owned by the
-     *  WriteControl
+     *  shared const pointer to Container.
      */
-    const Container* getContainer()
+    mem::SharedPtr<const Container> getContainer()
```
Comment (asylvest) [inline six/modules/c++/six/include/six/WriteControl.h]: You're providing a `const` container so the method itself can logically be `const` right?

Comment (asylvest) [inline six/modules/c++/six/include/six/WriteControl.h]: Since you have a non-const version in `ReadControl`, might as well put that version in too

**Comment by asylvest** [comment]
Just that one little thing to fix and then this is good to go

**Comment by JonathanMeans** [comment]
It’s fixed.

**Comment by asylvest** [comment]
This is good to go but there are merge conflicts... let me know once they're resolved and I'll merge it

**Comment by JonathanMeans** [comment]
This is ready
